### ISSUE TITLE
[Question] how to really disable TLS protocols lower than 1.2 in TLS client?

### ISSUE THREAD

**Issue by pgp** [randombit/botan#2935]
Hi, I am implementing a curl-like http client, and I noticed that SSLLabs client test at:
https://clienttest.ssllabs.com:8443/ssltest/viewMyClient.html
shows as TLS client-supported versions, SSL3, TLS1.0 and TLS1.1, while I thought it shouldn't.
When the `Botan::TLS::Client` object is created [here](https://github.com/pgp/XFilesRootHelper/blob/httpsClient/RootHelper/tls/botan_rh_tls_descriptor.h#L346), I set explicitly as minimum version 1.2
Moreover, the [amalgamated source code](https://github.com/pgp/XFilesRootHelper/blob/httpsClient/genBotanAmalgamations.sh#L60) is created by specifying `--disable-modules=pkcs11,tls_10`.
The SSLLabs report (plain html) is attached:
[viewMyClient.zip](https://github.com/randombit/botan/files/8245517/viewMyClient.zip)
and can be generated running the cli command:
`./r httpsd https://clienttest.ssllabs.com:8443/ssltest/viewMyClient.html`

Am I missing something? Should I specify some additional parameter when building the tls client object, in order to allow TLS 1.2 only?
I'm using Botan v.2.18.0

**Comment by randombit** [comment]
This seems weird. Can you wireshark the exchange and attach the `pcap`?

**Comment by pgp** [comment]
Here it is, together with a stdout transcript of the command, and the html report file:
[wireshark_tcpdump.zip](https://github.com/randombit/botan/files/8251664/wireshark_tcpdump.zip)
e

**Comment by hrantzsch** [comment]
TL;DR: You should be fine, your client does not support version TLS 1.0. However, in TLS 1.2 the server cannot detect the list of supported versions by the initial connection attempt.

Negotiating versions in TLS 1.2 seems to be a bit complicated. When sending the initial client hello, there are two version fields involved:
 * the [version field in the outer "record layer"](https://datatracker.ietf.org/doc/html/rfc5246/#section-6.2.1) (`TLSPlaintext.version`)
 * the [client_version field in the inner "handshake layer"](https://datatracker.ietf.org/doc/html/rfc5246/#section-7.4.1.2) (`ClientHello.client_version`)

The client hello in your pcap signals version 1.0 in the record layer and version 1.2 in the handshake layer. As far as I can see in the RFC, it is not clearly specified how this has to be interpreted by the server. Apparently, it _could_ be [interpreted as the lowest version supported by the client is TLS 1.0](https://datatracker.ietf.org/doc/html/rfc5246/#appendix-E.1), but this reads more like a suggestion to me.

The real negotiation of the version happens only after the server hello has been received. If the server does not support TLS 1.2, it will specify a lower version here. The client will then either downgrade to that version, if supported, or abort the handshake. In your case it should abort the handshake because you specified that TLS lower than 1.2 should not be used.

Note that Botan 2.* the [record layer version will always appear as TLS 1.0](https://github.com/randombit/botan/blob/3d5a0ade84a6db492a493e8872e024a27695556e/src/lib/tls/tls_handshake_io.cpp#L47). This has been changed on the current master branch, which will be released as Botan 3.0 at some point.

**Comment by hrantzsch** [comment]
I'm not sure how exactly ssllabs tries to determine the supported versions in a TLS 1.2 client. Like said above, they can only try to downgrade to, e.g., TLS 1.1 in the server hello and see how the client reacts.

You can actually see that `curl --tlsv1.2 'https://clienttest.ssllabs.com:8443/ssltest/viewMyClient.html'` produces the same results, where versions older than TLS 1.2 seem to be supported, although they are not.
